src/voter/voting_round.rs

```diff
@@ -232,12 +233,45 @@ impl<H, N, E: Environment<H, N>> VotingRound<H, N, E> where
 						self.env.precommit_equivocation(self.votes.number(), e);
 					}
 				}
+				Message::Primary(primary) => {
+					let primary_index = self.votes.number() as usize % self.votes.voters().voters.len();
```

Comment:
even though this has access to `voters` I would prefer some APIs rather than direct field access